### PR TITLE
feat: Implemented Endpoint to add a new user to the Waitlist

### PR DESCRIPTION
**Pull request by Nifilat**
​

## Description

The waitlist endpoint was designed to manage user registrations and store user details for the product until it gets launched. This feature allows users to reserve their spot in line for future updates or access to the boilerplate. Key aspects of the implementation include:

- Users can sign up by providing their email address and full name
- There is validation to check for duplicate emails or if the name field is null
- A confirmation email is sent to the email address provided upon successful sign-up
- To protect against abuse, the waitlist endpoint incorporates rate-limiting

​

## Related Issue (Link to issue ticket)
[WAITLIST API-BACKEND](https://github.com/hngprojects/hng_boilerplate_python_fastapi_web/issues/21)

​

## Motivation and Context

This ensures that users on the waitlist can have first-hand information about new features and when the product finally launches.

​

## How Has This Been Tested?

I have tested locally using both Postman and unit tests that I wrote
​

### Success Response
Screenshot 2024-07-22 at 07 53 53

### Error Response
Screenshot 2024-07-22 at 07 54 04

Screenshot 2024-07-22 at 07 53 28

​

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
      ​

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation as needed.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: .env.sample [hunk 1]
```diff
@@ -14,8 +14,8 @@ ACCESS_TOKEN_EXPIRE_MINUTES = 10
...
-MAIL_USERNAME=""
-MAIL_PASSWORD=""
-MAIL_FROM=""
+MAIL_USERNAME=''
+MAIL_PASSWORD=''
+MAIL_FROM=''
 MAIL_PORT=465
-MAIL_SERVER="smtp.gmail.com"
\ No newline at end of file
+MAIL_SERVER='smtp.gmail.com'
```
Comment (Goketech) [inline .env.sample:21]: revert all changes in this file

File: tests/v1/test_waitlist.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+# Load environment variables
+load_dotenv()
+
+# Database setup for testing
+SQLALCHEMY_DATABASE_URL = config('DB_URL')
+engine = create_engine(SQLALCHEMY_DATABASE_URL)
+TestingSessionLocal = sessionmaker(
+    autocommit=False, autoflush=False, bind=engine)
+
```
Comment (Goketech) [inline tests/v1/test_waitlist.py:23]: mock db for testing

**Review by Goketech — CHANGES_REQUESTED** [review]
pull from dev branch, follow the file structure and make changes

File: api/v1/routes/waitlist.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+from fastapi import APIRouter, HTTPException, Depends, Request
+from sqlalchemy.orm import Session
+from pydantic import BaseModel, EmailStr
+from api.v1.models.user import WaitlistUser
+from api.v1.schemas.waitlist import WaitlistUserCreate
+from api.db.database import get_db
+from api.utils.rate_limiter import rate_limit
+from api.v1.services.email.waitlist_email import send_confirmation_email
+
+router = APIRouter()
```
Comment (joboy-dev) [inline api/v1/routes/waitlist.py:10]: Use prefix here ans set it to /waitlist

File: api/v1/services/email/waitlist_email.py
Comment (joboy-dev) [inline api/v1/services/email/waitlist_email.py:1]: There is a function handling emails

**Review by joboy-dev — CHANGES_REQUESTED** [review]
Make changes

**Comment by Nifilat** [comment]
This is an old PR and can be closed. Someone already worked on this feature using the newly appointed spreadsheet method.On 27 Jul 2024, at 12:25, Adegbehingbe, Oluwakorede Joseph ***@***.***> wrote:﻿
@joboy-dev requested changes on this pull request.

Make changes

**Pull request closed without merging**